### PR TITLE
feat: allow individual extension configs

### PR DESCRIPTION
**Pull request by leoluz**
Checklist:

* [ ] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [ ] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [ ] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [ ] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [ ] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [ ] I have added a brief description of why this PR is necessary and/or what this PR solves.
* [ ] Optional. My organization is added to USERS.md.
* [ ] Optional. For bug fixes, I've indicated what older releases this fix should be cherry-picked into (this may or may not happen depending on risk/complexity).

### CODE REVIEW

File: server/extension/extension.go [hunk 1]
```diff
@@ -410,28 +410,46 @@ func proxyKey(extName, cName, cServer string) ProxyKey {
...
+			backendConfig := BackendConfig{}
+			err = yaml.Unmarshal(parsedExtConfigBytes, &backendConfig)
+			if err != nil {
+				return nil, fmt.Errorf("invalid parsed backend extension config for extension %s: %w", extName, err)
+			}
+			ext := ExtensionConfig{
+				Name:    extName,
+				Backend: backendConfig,
+			}
+			configs.Extensions = append(configs.Extensions, ext)
```
Comment (crenshaw-dev) [inline server/extension/extension.go:449]: I think, as written, this function can produce two extensions with the same name. We should probably dedupe here.

**Review by crenshaw-dev — CHANGES_REQUESTED** [review]

File: docs/developer-guide/extensions/proxy-extensions.md [hunk 2]
```diff
@@ -63,6 +63,33 @@ data:
...
+above using a dedicated key:
+
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: argocd-cm
+  namespace: argocd
+data:
+  extension.config.httpbing: |
```
Comment (crenshaw-dev) [inline docs/developer-guide/extensions/proxy-extensions.md:78]:
```suggestion
  extension.config.httpbin: |
```

File: util/settings/settings.go [hunk 3]
```diff
@@ -125,7 +125,7 @@ type ArgoCDSettings struct {
 	AppsInAnyNamespaceEnabled bool `json:"appsInAnyNamespaceEnabled"`
 	// ExtensionConfig configurations related to ArgoCD proxy extensions. The value
 	// is a yaml string defined in extension.ExtensionConfigs struct.
```
Comment (crenshaw-dev) [inline util/settings/settings.go:127]: Need to update comment

File: server/server.go [hunk 4]
```diff
@@ -706,7 +706,7 @@ func (a *ArgoCDServer) watchSettings() {
 			log.Infof("gogs secret modified. restarting")
 			break
 		}
-		if prevExtConfig != a.settings.ExtensionConfig {
+		if !equalExtConfig(prevExtConfig, a.settings.ExtensionConfig) {
```
Comment (crenshaw-dev) [inline server/server.go:709]: Maybe just reflect.DeepEqual? Less code for us, their implementation looks pretty simple. Don't think it's performance-critical code.

File: docs/developer-guide/extensions/proxy-extensions.md [hunk 5]
```diff
@@ -63,6 +63,37 @@ data:
...
+    - url: http://httpbin.org
+      headers:
+      - name: some-header
+        value: '$some.argocd.secret.key'
+      cluster:
+        name: some-cluster
+        server: https://some-cluster
+```
+
+Attention: Extension names must unique in Argo CD configmap. If
```
Comment (crenshaw-dev) [inline docs/developer-guide/extensions/proxy-extensions.md:93]:
```suggestion
Attention: Extension names must be unique in the Argo CD configmap. If
```

File: docs/developer-guide/extensions/proxy-extensions.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (leoluz) [reply docs/developer-guide/extensions/proxy-extensions.md:78]: updated the docs describing the behaviour.

File: docs/developer-guide/extensions/proxy-extensions.md [hunk 6]
```diff
@@ -63,6 +63,37 @@ data:
...
+duplicated keys are found, Argo CD API server will log an error
```
Comment (crenshaw-dev) [inline docs/developer-guide/extensions/proxy-extensions.md:94]:
```suggestion
duplicated keys are found, the Argo CD API server will log an error
```

File: server/extension/extension.go (hunk 1 above)
Comment (leoluz) [reply server/extension/extension.go:449]: duplicated extension names aren't allowed by the validation logic. updated docs

**Review by crenshaw-dev — CHANGES_REQUESTED** [review]

File: server/server.go (hunk 4 above)
Comment (leoluz) [reply server/server.go:709]: fixed

File: docs/developer-guide/extensions/proxy-extensions.md (hunk 5 above)
Comment (leoluz) [reply docs/developer-guide/extensions/proxy-extensions.md:93]: fixed

File: docs/developer-guide/extensions/proxy-extensions.md (hunk 6 above)
Comment (leoluz) [reply docs/developer-guide/extensions/proxy-extensions.md:94]: fixed

File: util/settings/settings.go (hunk 3 above)
Comment (leoluz) [reply util/settings/settings.go:127]: done

File: docs/developer-guide/extensions/proxy-extensions.md (hunk 2 above)
Comment (leoluz) [reply docs/developer-guide/extensions/proxy-extensions.md:78]: fixed

**Review by crenshaw-dev — APPROVED** [review]

**Pull request merged**